File: src/150redo/Trees/go/98.go
package trees

import "math"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	return f(root, math.MaxInt, math.MinInt)
}

func f(curr *TreeNode, maxVal int, minVal int) bool {
	if curr == nil {
		return true
	}

	if curr.Val >= maxVal || curr.Val <= minVal {
		return false
	}

	return f(curr.Left, curr.Val, minVal) && f(curr.Right, maxVal, curr.Val)
}

File: src/2025/XPrep/98.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  public:
    bool isValidBST(TreeNode *root) {
        // at every recursive step, maintain a max value on left and min value on right
        if (!root) return true;
        return f(root->left, LONG_MIN, root->val) && f(root->right, root->val, LONG_MAX);
    }

    bool f(TreeNode *curr, long L, long R) {
        if (!curr) return true;
        if (!(curr->val > L && curr->val < R)) return false;
        return f(curr->left, L, curr->val) && f(curr->right, curr->val, R);
    }
};